### PR TITLE
Assorted cleanups II

### PR DESCRIPTION
**Pull request by fstirlitz**
<details> <summary> Boilerplate (own code, improvement) </summary>

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)

### What is the purpose of your *pull request*?
- Improvement

</details>

(Spritual follow-up to #372.)

Most of these commits remove dead code and compatibility hacks meant for no longer supported Python versions. The SWF interpreter is removed again; it was pretty silly to keep it around in the first place.

The patches are pretty much unrelated to each other, so feel free to cherry-pick them individually as they are pushed on this branch.

Some further cleanups to consider:

- ~~`_windows_write_string` in utils seems to be a workaround for <https://bugs.python.org/issue1602>, which is apparently fixed in 3.6.  We should be able to get rid of it too. (I’m not sure what `write_string` itself adds over plain `sys.stderr.write` either.)~~
- ~~`locked_file` may be re-written as a `@contextlib.contextmanager`~~
- `uppercase_escape` and `lowercase_escape`: those functions seem as bad an idea as `addslashes` and `stripslashes` in PHP. The former doesn’t appear to be used anywhere at all other than in tests; the latter has a number of appearances in extractors, most of which seem pretty redundant, e.g.

   https://github.com/yt-dlp/yt-dlp/blob/11aa91a12f95821500fa064402a3e2c046b072fb/yt_dlp/extractor/chaturbate.py#L47-L51

   where the JSON parser is supposed to take case of `\uXXXX` escapes already.
- Split `utils` into thematically-linked modules. Right now it contains functions for as disparate subtitle conversion, HTML parsing, dictionary traversal, bug workarounds, command line handling… It’s become an unmanageable mess.
- Likewise, move out ‘common’ extractor methods from `InfoExtractor`, which has also become something of a god object anti-pattern. (Say, move manifest parsing into its own modules, one per manifest format.)

### CODE REVIEW

**Comment by fstirlitz** [comment]
I think I am done for now, but leaving this as draft because this should still not be squashed; each commit should be individually reviewed and cherry-picked. (Actually, I have a patch as well that gets rid of `compat_str`, `compat_chr` and `compat_ord`, but it’s too invasive to comfortably include here.)

**Comment by pukkandan** [comment]
> uppercase_escape and lowercase_escape: those functions seem as bad an idea as addslashes and stripslashes in PHP. The former doesn’t appear to be used anywhere at all other than in tests; the latter has a number of appearances in extractors, most of which seem pretty redundant, 
where the JSON parser is supposed to take case of \uXXXX escapes already.

It is for handling double escaping

> Split utils into thematically-linked modules. Right now it contains functions for as disparate subtitle conversion, HTML parsing, dictionary traversal, bug workarounds, command line handling… It’s become an unmanageable mess.

This is a good idea. We can make utils a directory with relevent submodules. By importing all the submodules in it's `__init__`, we can keep doing `from .utils import ...` while keeping the code more managable

> Likewise, move out ‘common’ extractor methods from InfoExtractor, which has also become something of a god object anti-pattern. (Say, move manifest parsing into its own modules, one per manifest format.)

This is harder. There is an annoying amount of coupling b/w the various parts. Without serious refactoring, we will need to pass around either an `ie` or `ydl` object to all the modules (similar to logger situation in cookies.py, but worse)

> An idea I just had: turn compat into a directory, move the current contents into __init__.py and create a file re.py with the following contents:

Another smart idea. I'm all for it

---

I dont want to remove the older functions right now though. There are some other parts that I have been meaning to change/remove (see the `DeprecationWarning`s). We should add similar warnings to the to-be-removed functions for now and only remove it after a proper announcement

**Comment by fstirlitz** [comment]
To keep compatibility with upstream pull requests? Right, I can see the concern.

(Besides, `DeprecationWarning` is an actual class in Python, supposed to be used in conjunction with the `warnings` module, not just a fancy label to print…)

> It is for handling double escaping

The example I cited looks like it could be solved by performing `_parse_json` twice. In fact, it should be more robust that way. The others I have not looked at in much depth, though.

> There is an annoying amount of coupling b/w the various parts. Without serious refactoring, we will need to pass around either an ie or ydl object to all the modules

For this reason I kind of regret having suggested 3.6 as the minimum supported Python version instead of 3.7, because the latter implements `contextvars` ([PEP 567](https://www.python.org/dev/peps/pep-0567/)), which I think would be of tremendous help during the refactoring, but shimming this API correctly for 3.6 is not easy in the presence of asynchronous code (which we don’t have much of *yet*, but…).

**Comment by nyuszika7h** [comment]
What upstream though? It's becoming pretty clear that the youtube-dl maintainers are no longer interested in maintaining the project beyond keeping YouTube downloads minimally working (still throttled). When YouTube breaks entirely, dstftw wakes up to push a quick fix and then goes back to hibernation.

**Comment by pukkandan** [comment]
> Besides, DeprecationWarning is an actual class in Python, supposed to be used in conjunction with the warnings module, not just a fancy label to print…

😄 The name being same is co-incidental. My intend was only to print a message. But perhaps it makes sense to use the warnings module instead 🤔

**Comment by fstirlitz** [comment]
The upstream maintainers may be asleep at the wheel, but people who haven’t gotten the memo yet (or just stubbornly refuse to switch) are still filing pull requests against the original repository. Maintaining upstream-compatible interfaces would ease porting those PRs into our codebase, often making copy-paste sufficient. On the other hand, assuming that upstream is truly dead and will never merge them, there is less concern over merge conflicts when porting later upstream-submitted patches; when porting extractor PRs we can adjust code structure once to our standards and maintain them in that form, knowing we will not have to port further upstream patches. It’s no longer a clear win to maintain those compatible interfaces if we can instead fix up extractors as we import them.

> But perhaps it makes sense to use the warnings module instead

Given that using the name might mislead people into thinking they can e.g. use the `warnings` module to suppress the warning, yes, probably.

**Comment by fstirlitz** [comment]
> By importing all the submodules in it's `__init__`, we can keep doing `from .utils import ...` while keeping the code more managable

Actually, part of the motivation to split `utils` is to reduce the module footprint, so that extractors only have to load what they will end up using, instead of everything at once. (Especially that most of the time, extractors only end up using a handful of functions.) Lumping everything under the `utils` module and keeping modules importing from there defeats that. I’d expect the `utils` imports to only be kept for backwards compatibility.

**Review by fstirlitz — COMMENTED** [review]
b8d521f02d472f9a43408556a30035fc71a38426

Should we not set up a minimal business-card website on GHP, like the upstream used to have (before they moved to Uberspace)? Would those scripts be useful there?

**Comment by pukkandan** [comment]
? Should we not set up a minimal business-card website on GHP, like the upstream used to have (before they moved to Uberspace)? Would those scripts be useful there?

Those scripts seemed very specific to the exact setup youtube-dl had. I can revert it if you think they would be useful. But they are always saved in the git history anyway 🤷 

Speaking of GHP, I have zero experience with it. So I wont be trying to make one, atleast in the forseeable future. There was some (private) discussion about it before, but no one has been willing to put in the effort so far.

---

> By importing all the submodules in it's __init__, we can keep doing from .utils import ... while keeping the code more managable

> Actually, part of the motivation to split utils is to reduce the module footprint, so that extractors only have to load what they will end up using, instead of everything at once. (Especially that most of the time, extractors only end up using a handful of functions.) Lumping everything under the utils module and keeping modules importing from there defeats that. I’d expect the utils imports to only be kept for backwards compatibility.

If we don't import everything in `utils.__init__`, backward compatibility will be completely broken. If we do, all functions will be loaded even when only a submodule is accessed. The only way I can think of achieving both at the same time is to use [module level dunders](https://docs.python.org/3/reference/datamodel.html#customizing-module-attribute-access)

**Comment by fstirlitz** [comment]
Of course, code written against `utils` has to keep working. The point is not to _have to_ import everything through `utils`, which will in turn pull in loads of other unrelated code, but import smaller modules with only the required functionality.

As for GHP, I don’t know. The old setup could be replicated here for the time being.

**Comment by fstirlitz** [comment]
I had to excise `locked_file` conversion from this patch series because it conflicts with #1066.

I think `sanitize_open` should be converted into a context manager as well, but that I expect to be more invasive.

File: yt_dlp/YoutubeDL.py [hunk 1]
```diff
@@ -911,10 +911,8 @@ def report_warning(self, message, only_once=False):
             self.to_stderr(f'{self._format_err("WARNING:", self.Styles.WARNING)} {message}', only_once)
 
     def deprecation_warning(self, message):
-        if self.params.get('logger') is not None:
-            self.params['logger'].warning('DeprecationWarning: {message}')
-        else:
-            self.to_stderr(f'{self._format_err("DeprecationWarning:", self.Styles.ERROR)} {message}', True)
+        import warnings
```
Comment (pukkandan) [inline yt_dlp/YoutubeDL.py:914]: any reason why the import is not at module level?

Also, I am not sure how warnings module works with the windows exe. There are a couple of places where the deprecation warnings are user-facing (as opposed to API only). So I'll need to test that

Comment (fstirlitz) [reply yt_dlp/YoutubeDL.py:914]: Optimizing for the happy path, where this never gets invoked? Also, putting imports at the top creates the potential for merge conflicts.

Do we catch deprecation warnings in CI? I think at least the 3.10 run should have warnings converted into errors.

**Comment by fstirlitz** [comment]
I think some of those are ready and can be cherry-picked as they are:

- `.compat` conversion: 9ed27e9690746f9f05dc6e37a31ef526f173186d, d4be2a367cbef3da2d20cbada17aebb1c7217fe6, c45f94c6c70fb9581defe3641e2b0e8f1cb3f817
- 96a9e84287661cbf62c91c7427e1ece6d7e77c8c
- 2dda4b0a255ca8209d81589e40efa6b30be53f04
- ea71883d47e2e0180a26b29eb73a38161b1ff344
- 16292448303fe1b91850c194bdb5ebb39fc0de5f
- Windows hacks: 30fa1bc19739d50be8f82b0ec022215413ded5a0 and 7b8b5303488321c6d625d2b216e2808ea2912830

Even a couple of removals are uncontroversial, I think: d020a00f6f4ce18c725ab73d5f7ebef979ea73b7, a337317fc8374a5e6d0d11207ca40a0a15d17157 and 1283f6f49cbb2634d49c413f1fbab4f7722fcbd9, since there is virtually no chance any of those will be used in extractors merged from upstream in the future.

Is anything stalling those in particular?

**Comment by pukkandan** [comment]
Most of these are good (except the compat definitions should still be kept imo). What is stalling this is only my lack of time and that this is a relatively low priority PR. I'll try to get atleast some of this merged before the next release.

**Comment by coletdjnz** [comment]
Noticed this branch the other day, might be something else we can remove?
https://github.com/yt-dlp/yt-dlp/blob/e7870111e83033e0ac728d5a2d565d1eb146c335/yt_dlp/utils.py#L1311-L1321

**Comment by pukkandan** [comment]
> create a file re.py with the following contents:

What do you think of directly patching the system modules instead? Since we are only adding attributes, and not editing any existing ones, doing this should be fine imo. Is this generally considered bad practice?

PS: During the last convo, I didn't remember this option at all, despite already doing this for `asyncio.run`

> I had to excise locked_file conversion from this patch series because it conflicts with https://github.com/yt-dlp/yt-dlp/pull/1066.

Could you give me a link to the commit (finding it is difficult since u force-pushed). I need to make some changes to file locking and can probably incorporate your changes when doing that

**Comment by fstirlitz** [comment]
> What do you think of directly patching the system modules instead?

- Our shims may contain bugs, be incomplete, or slightly deviate from original CPython implementations
- Monkey-patching system modules may affect other users of the same modules, thus generating spoopy action-at-a-distance; completely unrelated modules may work or not depending on whether `yt_dlp` is imported. This seems bad.

If we keep shims in interface-compatible modules that only we import, only we are going to be affected by potential bugs and subtle divergences.

It’s not that much different from mutating built-in objects’ prototypes in JavaScript. Doing it to backport features can be kind of okay, but only if you have full control over the entire execution environment; and unless you are sure your polyfills are 100% correct and compatible, better not.

> Could you give me a link to the commit

One version is in d1ac5dfe9432077747df0da7719847d5f10faea2. I don’t remember preparing anything significantly different from that.

**Comment by fstirlitz** [comment]
> can probably incorporate your changes when doing that

Didn’t make it, I suppose?

**Comment by pukkandan** [comment]
Because of the way the `FileDownloader`s are written, we need `locked_file(...).close()` to close the file and exit the context manager. I couldn't figure out how that can be done with `contextlib.contextmanager` (afaik it's not possible)

**Comment by pukkandan** [comment]
I would like a second opinion on https://github.com/yt-dlp/yt-dlp/pull/2173/commits/503c62cd364b4d1d99bddc3c9d86ed89e0365708. I don't particularly like having to mess with environment variables, but being able to tell whether or not the code is running from CLI or API is usefull. We could use this info to modify the `bug_reports_message` as well

**Comment by fstirlitz** [comment]
https://docs.python.org/3.7/library/contextvars.html

It doesn't look too hard to propagate manually either. Pass it to `main`, which should pass it to the constructor of `YoutubeDL`.

**Comment by fstirlitz** [comment]
But honestly, ideally warnings directed at programmers and those directed at CLI users would use completely different codepaths.

**Comment by pukkandan** [comment]
1. Deprecated **features** should act like `report_warning` when used through **CLI**
1. Deprecated **function calls** should also display a warning on **CLI** (maybe with added traceback?) so that the issue is noticed asap
1. Deprecated **function calls** when used though API should use `warnings.warn`
1. I'm not sure how deprecated **features** should show on **API**. Either method seems fine for this

Does this make sense?

**Comment by fstirlitz** [comment]
On second thought, the `warnings` module doesn’t seem all too well-designed itself. Perhaps the `DeprecationWarning` stuff should be discussed separately, but I don’t particularly feel like opening another ticket for it…

On an unrelated note, anything standing in the way of `compat` conversion into a directory? I thought that would be an easy one.

**Comment by pukkandan** [comment]
I have pushed some more changes. They are not finalized, if you have any opinions on them, I'd like to hear it. (Ignore https://github.com/yt-dlp/yt-dlp/pull/2173/commits/39bb6982c8f6324e38a7fdb33c089ab90e8f855f for now)

A particularly important commit is https://github.com/yt-dlp/yt-dlp/pull/2173/commits/7217ae4b377868b6037637fc8bc3e72fece1dfb9. This allows us to show deprecation warnings for compat functions without actually fully removing them. While this is arguably not very useful for compat, the same mechanism can be used to split up utils into sub-modules (while maintaining compatibility)

As for actually removing the compat/util functions - I want to remove most of the deprecated functions in one go after making an announcement to that effect. This is the main reason I have been adding deprecation warnings throughout the code and the reason I want to make sure these warnings are prominently visible.

**Comment by fstirlitz** [comment]
I’d name the module something like `_legacy`, but otherwise it’s a pretty good idea. Even better would be to:

- move even those deprecated symbols that are still used into `_legacy`, but also add [a warning filter](https://docs.python.org/3/library/warnings.html#describing-warning-filters-1) that suppresses the warnings in specific modules (or even lines)
- set up warnings to fail CI (`-bb -Werror`)
- successively remove items from the suppression list

In other words, practise [Shitlist-Driven Development](https://sirupsen.com/shitlists).

File: yt_dlp/postprocessor/metadataparser.py [hunk 2]
```diff
@@ -1,29 +1,25 @@
 import re
-from enum import Enum
 
 from .common import PostProcessor
+from ..utils import Namespace
 
 
 class MetadataParserPP(PostProcessor):
-    class Actions(Enum):
```
Comment (fstirlitz) [inline yt_dlp/postprocessor/metadataparser.py:8]: Actually, this one was fine. The fact that doing `assert isinstance(action, self.Actions)`, as you did later, makes sense makes it a proper type.

Comment (pukkandan) [reply yt_dlp/postprocessor/metadataparser.py:8]: I never liked how the function was decided using `getattr(self, action.value)`

File: yt_dlp/downloader/common.py [hunk 3]
```diff
@@ -72,12 +72,30 @@ class FileDownloader:
...
+            'deprecation_warning',
+            'report_error',
+            'report_file_already_downloaded',
+            'report_warning',
+            'to_console_title',
+            'to_stderr',
+            'trouble',
+            'write_debug',
+        ):
+            setattr(self, func, getattr(ydl, func))
```
Comment (fstirlitz) [inline yt_dlp/downloader/common.py:94]: I think the best way to do logging would be to split off the logging functionality into a separate object that only deals with logging and nothing else, and just pass that object around to whoever needs it. And then do away with the mirroring. It would be very easy to adapt `yt_dlp.cookies` to this style at least.

Similar strategy could be applied to network requests; I believe @coletdjnz’s draft PRs follow that route. We do this enough and we’ll discover that `YoutubeDL` has basically no responsibilities other than being a grab-bag of other objects…

Comment (coletdjnz) [reply yt_dlp/downloader/common.py:94]: > Similar strategy could be applied to network requests; I believe @coletdjnz’s draft PRs follow that route.

Yeah... I literally just copied in these 'mirroring' functions from FD... I agree it would be better to pass around some sort of logging object (or some global one that is imported?). 

There is the official python logging library; is there any reason we don't use that with our own custom logger?

Comment (fstirlitz) [reply yt_dlp/downloader/common.py:94]: I’m not sure the built-in `logging` would necessarily fit yt-dlp very well.

yt-dlp two kinds of messages: event messages and progress messages. Event messages are one-shot and permanent; progress messages are ephemeral and re-written to reflect current state. `logging` seems to be built to only handle event messages.

If we get parallel downloading some day, we would probably like for each downloader to display its own progress bar, while allowing potential intervening log messages to show as well without corrupting the display. I don’t think `logging` was built to handle such situations.

File: devscripts/run_tests.sh [hunk 4]
```diff
@@ -11,4 +11,4 @@ else
     exit 1
 fi
 
-python3 -m pytest "$test_set"
+python3 -bb -Werror -m pytest "$test_set"
```
Comment (pukkandan) [inline devscripts/run_tests.sh:14]: do same in devscripts\run_tests.bat

Comment (fstirlitz) [reply devscripts/run_tests.sh:14]: The reason I didn’t is that `run_tests.bat` invokes a `pytest` binary, which I assume takes the options meant for the module, not for the Python interpreter itself (those that come after `-m`).

I suppose I might try enabling the option with environment variables instead…

Comment (pukkandan) [reply devscripts/run_tests.sh:14]: You can replace it with `py -bb -Werror -m pytest`

Comment (pukkandan) [reply devscripts/run_tests.sh:14]: nvm, `py` may not always exist. Should be `python`. But then it can technically be python 2 :(

Comment (fstirlitz) [reply devscripts/run_tests.sh:14]: Could have used `python3` inste—

> This PEP deliberately excludes any proposals relating to Microsoft Windows, as devising an equivalent solution for Windows was deemed too complex to handle here.

— <https://peps.python.org/pep-0394/#exclusion-of-ms-windows>

Nevermind.

Comment (nyuszika7h) [reply devscripts/run_tests.sh:14]: Could just use `py -3` for Windows.

**Comment by pukkandan** [comment]
I have merged most of this PR now. What is remaining is:
- [ ] 1. Handling of `deprecation_warning`
- [ ] 2. Removing things from utils - Not planning to remove them atm. See https://github.com/yt-dlp/yt-dlp/pull/2173#issuecomment-1097387226
- [ ] 3. Adding warning for more compat options (https://github.com/yt-dlp/yt-dlp/pull/2173#issuecomment-1097711248) - I don't think this would be easy to implement, or very useful for that matter. It would be better if we can write some test that runs on the commit/PR **diff**s and check that the deprecated functions are not being used. This would allow us to strictly enforce the deprecation on all newly added code without messing with the 1500+ extractors
- [ ] 4. Separating out logger (https://github.com/yt-dlp/yt-dlp/pull/2173#discussion_r849266105) - This would need too many changes for us to seriously consider as part of this PR. Perhaps eventually...

which leaves us with deprecation_warning. This is roughly how I am thinking of handling it. Any issues?

```py
class YoutubeDL:
    ...
    def deprecated_feature(self, message):
        self.to_stderr(f'Deprecated Feature: {message}', only_once=True)

    def deprecation_warning(self, message):
        deprecation_warning(message, printer=self.report_error)

# utils.py
def deprecation_warning(message, printer=None):
    if IN_CLI:
        return (printer or write_string)(f'ERROR: {message}{bug_reports_message()}\n', only_once=True)
    else:
        import warnings
        warnings.warn(DeprecationWarning(message))
```
